### PR TITLE
fix retry cancellation when a batch fails on call attempt

### PR DESCRIPTION
**Pull request by markdroth**
I ran into this problem while diagnosing the timeout in grpc_tool_test in #26766.

First, some background: The retry code sometimes needs to defer callbacks to the surface.  For example, if it sees an error in on_complete for send_initial_metadata, it cannot deliver that callback to the surface until it has seen recv_trailing_metadata_ready, because if it is going to retry the call, then it would not be appropriate to report the failure for send_initial_metadata to the surface, since that op will be replayed on the next call attempt, where it may succeed.  Note that when this happens, if the application has not yet started the recv_trailing_metadata op, the retry code will actually start that op on its own internally, and the result of that op will be cached until the application starts the op later.  This is necessary because the retry code needs to know the call attempt's status (which is delivered with recv_trailing_metadata_ready) in order to decide whether or not to retry the call.

In the test failure I was seeing, it turned out that what was happening is that the test was starting the send_initial_metadata op, which triggered the LB pick, which then failed, because there was no server running.  However, the test had not yet started the recv_trailing_metadata op, so when the retry code saw the failure of the send_initial_metadata op, it started its own recv_trailing_metadata op internally.  But it failed to cancel the call attempt before doing that, so the LoadBalancedCall did not know to fail the recv_trailing_metadata op -- instead, the LoadBalancedCall just queued the op because it didn't have a subchannel call and thought that it hadn't yet seen the send_initial_metadata op.

This PR fixes the retry code to cancel the call attempt whenever a batch fails.  This is what filters below it expect to see, since that's what the code in surface/call.cc does, but in the cases where the retry code defers a callback, the retry code is not giving the surface code the opportunity to do that, so it needs to start the cancellation itself.

### CODE REVIEW

File: test/core/end2end/tests/retry_lb_fail.cc [hunk 1]
```diff
@@ -0,0 +1,282 @@
+//
+// Copyright 2017 gRPC authors.
```
Comment (yashykt) [inline test/core/end2end/tests/retry_lb_fail.cc:2]: s/2017/2021
(looks like a new file to me)

**Comment by ctiller** [comment]
Medium term I'm hoping we just get rid of batches and simplify the codebase.

File: src/core/ext/filters/client_channel/retry_filter.cc [hunk 2]
```diff
@@ -376,21 +391,14 @@ class RetryFilter::CallData {
...
     // If server_pushback_md is non-null, sets *server_pushback_ms.
     bool ShouldRetry(absl::optional<grpc_status_code> status, bool is_lb_drop,
                      grpc_mdelem* server_pushback_md,
                      grpc_millis* server_pushback_ms);
 
-    // Cancels the call attempt.  Unrefs any deferred batches.
-    // Adds a batch to closures to cancel this call attempt.
-    void Cancel(const char* reason, CallCombinerClosureList* closures);
+    // Abandons the call attempt.  Unrefs any deferred batches.
+    void Abandon();
```
Comment (yashykt) [inline src/core/ext/filters/client_channel/retry_filter.cc:401]: I'm probably missing something, but why can't this just be the destructor?

**Review by yashykt — APPROVED** [review]

File: test/core/end2end/tests/retry_lb_fail.cc (hunk 1 above)
Comment (markdroth) [reply test/core/end2end/tests/retry_lb_fail.cc:2]: I copied this from the existing retry_lb_drop.cc file and just changed some of the details, so I retained the original date.

File: src/core/ext/filters/client_channel/retry_filter.cc (hunk 2 above)
Comment (markdroth) [reply src/core/ext/filters/client_channel/retry_filter.cc:401]: Because we don't know that we can destroy the object at the point at which it is abandoned, since there may still be other pending callbacks that need to complete on this attempt.

**Review by markdroth — COMMENTED** [review]
I don't think there's any reasonable way to completely proactively validate that we aren't missing edge cases here.  All we can really do is to add test cases as we encounter problems, which is what I'm doing here.  At this point, I'm actually relatively happy with the test coverage; I've found and eliminated a large number of edge cases over the last several months.  That is not an absolutely guarantee that there won't be additional problems once we ship this code, but I don't think there's any way to do that.

In the long term, I think that a lot of the state machinery necessary for the filter stack will go away as we move to Craig's new promise-based architecture.

Thanks for the review!

**Review by yashykt — APPROVED** [review]

**Comment by markdroth** [comment]
Known issues: #26595
